lib/src/browser_client.dart

```diff
@@ -41,20 +41,45 @@ class BrowserClient extends BaseClient {
 
   /// Sends an HTTP request and asynchronously returns the response.
   @override
-  Future<StreamedResponse> send(BaseRequest request) async {
+  Future<StreamedResponse> send(BaseRequest request, {Duration? timeout}) {
+    final completer = Completer<StreamedResponse>();
+    _send(request, timeout, completer);
+    return completer.future;
+  }
+
+  Future<void> _send(BaseRequest request, Duration? timeout,
+      Completer<StreamedResponse> completer) async {
+    Timer? timer;
+    late void Function() onTimeout;
+    if (timeout != null) {
+      timer = Timer(timeout, () {
+        onTimeout();
+      });
+      onTimeout = () {
+        completer.completeError(TimeoutException('Request aborted', timeout));
+      };
+    }
     var bytes = await request.finalize().toBytes();
     var xhr = HttpRequest();
     _xhrs.add(xhr);
+    unawaited(completer.future.whenComplete(() {
+      _xhrs.remove(xhr);
+    }));
     xhr
       ..open(request.method, '${request.url}', async: true)
       ..responseType = 'arraybuffer'
       ..withCredentials = withCredentials;
     request.headers.forEach(xhr.setRequestHeader);
-
-    var completer = Completer<StreamedResponse>();
+    if (timeout != null) {
+      onTimeout = () {
+        xhr.abort();
+        completer.completeError(TimeoutException('Request aborted', timeout));
+      };
+    }
 
     unawaited(xhr.onLoad.first.then((_) {
       var body = (xhr.response as ByteBuffer).asUint8List();
+      timer?.cancel();
       completer.complete(StreamedResponse(
```

Comment:
Should probably guard by an
```dart
  if (!completer.isCompleted) { 
    completer.complate(...)
  }
```
It's probably technically possible for the timer to have completed the completer, but the callback not having been called yet (it's not a sync completer!), and then an unguarded `completer.complete` here will throw a `StateError`. Basically a race condition.
Ditto below.


Or maybe it can't happen because the timeout is a timer event, so a top-level event loop event of its own, and it completing a future will happen in a microtask of the *same* event, the only microtask scheduled by that event, so nothing can get between them. Then that might be worth documenting.

Reply:
I went ahead and guarded them all with a `completer.isCompleted` checks in both client implementations. I think it's probably more checks than are needed, but adding the checks won't introduce bugs and we can always remove them later with an explanation of why it isn't required.